pymc3/distributions/multivariate.py

```diff
@@ -2048,3 +2050,100 @@ def random(self, point=None, size=None):
+        n = A.shape[0]
+        mu = np.zeros(n)
+        if A.ndim != 2 or n != A.shape[1] or np.linalg.norm(A - A.T, np.inf) > 1e-06:
+            raise ValueError("The adjacency matrix `A` needs to be a square and symmetric")
+        W = np.diag(A.sum(axis=0)) - A
+        cov = scipy.linalg.pinvh(W) / tau
+        return multivariate_normal.rng_fn(rng=rng, mean=mu, cov=cov, size=size)
+
+
+icar = ICARRV()
+
+
+class ICAR(Continuous):
+    r"""
+    The zero-sum constrained Intrinsic Conditional Autoregressive Model.
+
+    .. math::
+
+        f(x \mid \tau, A) =
+            (2 \pi \tau)^{(n - 1)/2}\sqrt{\prod_{i=1}^{n-1}d_i}
+            \exp\Big\{-\frac{\tau}{2}x^T H x\Big\}
+
+    where :math:`d_1 \geq d_2 \geq d_{n-1} \gt d_n = 0` are eigenvalues of
+    :math:`H` in decreasing order. :math:`H` is specified as a known matrix determined
+    by the neighborhood structure of the region of interest, such that
+
+    .. math::
+
+        (H)_{ij} = \left\{
+            \begin{array}{c l}
+                h_i & \text{if } \hspace{2mm} i = j\\
+                -g_{ij} & \text{if } \hspace{2mm} i \in N_{j}\\
+                0 & \text{otherwise},
+            \end{array}\right.
+
+    where :math:`g_{ij} \geq 0` is a measure of how similar subregions ``i`` and
+    ``j`` are, :math:`\{N_j;j=1,\dots,n\}` denotes the set of subregions that are
+    neighbors of subregion ``j``, :math:`g_{ij}=g_{ji}, and :math:`h_i=\sum_{i
+    \neq j}g_{ij}`.
+
+    ======== ================================================================================
+    Support   :math:`x \in \mathbb{R}^k`
+    Mean      :math:`0 \in \mathbb{R}^k`
+    Variance  :math:`\tau^{-1} H^{+}` where :math:`H^{+}` is the Moore–Penrose pseudoinverse.
+    ======== ================================================================================
+
+    Parameters
+    ----------
+    A: 2-d array/Tensor
+        Symmetric adjacency matrix of 1s and 0s indicating
+        adjacency between elements.
+    tau: float or 1-element array/tensor
+        Positive precision variable controlling the scale of the underlying normal variates.
+
+    References
+    ----------
+    .. Keefe, M. J., Ferreira, M. A. R., and Franck, C. T. (2018). “On the formal
+       specificationof sum-zero constrained intrinsic conditional autoregressive
+       models.” Spatial Statistics, 24: 54–65.
+       doi:https://doi.org/10.1016/j.spasta.2018.03.007.186,205
+
+    """
+    rv_op = icar
+
+    @classmethod
+    def dist(cls, A, tau, *args, **kwargs):
+        A = aesara.sparse.as_sparse_or_tensor_variable(floatX(A))
+        if aesara.sparse.basic._is_sparse_variable(A):
+            A = aesara.sparse.dense_from_sparse(floatX(A))
+        else:
+            A = at.as_tensor_variable(floatX(A))
+        tau = at.as_tensor_variable(floatX(tau))
+        Assert("`A` needs to be a 2D tensor")(at.eq(A.ndim, 2))
+        Assert("`tau` needs to be a 0D tensor")(at.eq(tau.ndim, 0))
```

Comment:
The output of these `Assert`s needs to be in the graph for them to take effect.

```suggestion
        A = Assert("`A` needs to be a 2D tensor")(A, at.eq(A.ndim, 2))
        tau = Assert("`tau` needs to be a 0D tensor")(tau, at.eq(tau.ndim, 0))
```